eth/backend.go

```diff
@@ -575,3 +578,31 @@ func (s *Ethereum) Stop() error {
 
 	return nil
 }
+
+func (s *Ethereum) HandleRequiredProtocolVersion(required params.ProtocolVersion) error {
+	var needLevel int
+	switch s.config.RollupHaltOnIncompatibleProtocolVersion {
+	case "major":
+		needLevel = 3
+	case "minor":
+		needLevel = 2
+	case "patch":
+		needLevel = 1
+	default:
+		return nil // do not consider halting if not configured to
+	}
+	haveLevel := 0
+	switch params.OPStackSupport.Compare(required) {
+	case params.OutdatedMajor:
+		haveLevel = 3
+	case params.OutdatedMinor:
+		haveLevel = 2
+	case params.OutdatedPatch:
+		haveLevel = 1
+	}
+	if haveLevel >= needLevel { // halt if we opted in to do so at this granularity
+		log.Error("opted to halt, unprepared for protocol change", "required", required, "local", params.OPStackSupport)
```

Comment:
nit: We're not particularly good at this, but log messages should start with a capital (error strings start with lower case as they're expected to be concatenated).